### PR TITLE
[native_assets_cli] Add `outputDirectoryShared` to config

### PR DESCRIPTION
**Pull request by dcharkes**
This PR adds an example of in-hook caching and sharing work across multiple hook invocations:

* https://github.com/dart-lang/native/issues/1375
* https://github.com/dart-lang/native/issues/1379

The prime example can be found in `pkgs/native_assets_builder/test_data/transformer/`.

In order for the example to work correctly, a new `BuildConfig.outputDirectoryShared` is added and locking is used to ensure that the same hook invoked twice concurrently with a different parameter doesn't cause concurrent modifications. (This should not regress Flutter builds, because these run hooks for the various targets in sequence.)

This CL does not yet address some other issues:

* Assets being used and hooks being run concurrently: https://github.com/dart-lang/native/issues/1534
* Using hashes of file contents for caching instead of file time stamps: https://github.com/dart-lang/native/pull/1533#issuecomment-2338477561
* A better API for directories that are partially shared: https://github.com/dart-lang/native/issues/1379#issuecomment-2348177710

### CODE REVIEW

**Review by HosseinYousefi — APPROVED** [review]
LGTM!

**Comment by mosuem** [comment]
Offline discussion with @dcharkes - do not submit until new version has rolled into flutter.

File: pkgs/native_assets_builder/test_data/transformer/hook/build.dart [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    final dataDirectory =
+        Directory.fromUri(config.packageRoot.resolve('data/'));
+    // If data are added, rerun hook.
+    output.addDependency(dataDirectory.uri);
+
+    var transformedFiles = 0;
+    var cachedFiles = 0;
+
+    await for (final sourceFile in dataDirectory.list()) {
+      if (sourceFile is! File) {
```
Comment (mosuem) [inline pkgs/native_assets_builder/test_data/transformer/hook/build.dart:21]: nit: filter with `whereType` instead

File: pkgs/native_assets_builder/test_data/transformer/hook/build.dart [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        continue;
+      }
+
+      final sourceName =
+          sourceFile.uri.pathSegments.lastWhere((e) => e.isNotEmpty);
```
Comment (mosuem) [inline pkgs/native_assets_builder/test_data/transformer/hook/build.dart:26]: In the old uri vs. Strings discussion - `package:path`  has nice methods to get the `basename` of a file

**Review by mosuem — APPROVED** [review]

File: pkgs/native_assets_builder/test_data/transformer/hook/build.dart (hunk 2 above)
Comment (dcharkes) [reply pkgs/native_assets_builder/test_data/transformer/hook/build.dart:26]: vs `package:file` 😄

**Pull request merged**